### DISCUSSION TITLE
🐞 dates w/ unix time will break after 2038

### DISCUSSION THREAD

**Discussion by grouville** [dagger/dagger#12399] [maintainers]
Following @sipsma's [comment](https://github.com/dagger/dagger/issues/4152#issuecomment-1344588891):
> Great catch, on a less urgent note we'll also run into problems representing dates w/ unix time after 2038 :-)

This follows a conversation with @JammUtkarsh, who wanted to contribute:
- https://github.com/dagger/dagger/issues/4152#issuecomment-1571930310

### Root cause: "Unix Millennium Bug."
Unix time is the number of seconds that have elapsed since January 1, 1970, at 00:00:00 UTC. It is typically stored as a 32-bit signed integer, which means it can represent dates up to 03:14:07 UTC on January 19, 2038. After this point, the 32-bit representation will overflow, and the time will wrap around to a negative value, causing incorrect date calculations.

### `dagger/graphql` library

The issue won't happen in our internal representation of Dates, in our graphQL library. As it seems to rely on the `time.Time` implementation, representation with 64 bytes. It was already tested: 
<img width="1406" alt="image" src="https://github.com/dagger/dagger/assets/31691250/1b7ee6c1-3c75-48b2-a545-59f7ac487896">

### Root cause

The root problem seems to be scoped to the use of `time.Unix()`: https://github.com/search?q=repo%3Adagger%2Fdagger+Unix%28&type=code

cc @JammUtkarsh

**Comment by jammutkarsh** [comment]
Please correct me if I am wrong, but this bug might be obsolete.

The fact that Unix systems uses 32-bit representation is a [known issue](https://en.wikipedia.org/wiki/Year_2038_problem).

The Go team was aware of this and is already implemented the `time` standard library with 64-bit representation. Specifically the [`Time.Unix()`](https://pkg.go.dev/time#Time.Unix) is already addressed.

Read More:
1. [Go source code](https://cs.opensource.google/go/go/+/refs/tags/go1.20.5:src/time/time.go;l=1192)
2. [A forum](https://forum.golangbridge.org/t/is-golang-2038-safe/14615)

**Comment by grouville** [comment]
> representation

@sipsma, just a bump as @JammUtkarsh is willing to contribute on this issue. I actually don't know what could be the issue ?
